src/hardware/vga_xga.cpp

```diff
@@ -912,20 +909,24 @@ void XGA_BlitRect(Bitu val) {
 			// Register (COLOR_CMP)" which this code holds as
 			// xga.color_compare.
 
-			if (xga.control1 & 0x100) {
-				// COLOR_CMP enabled. control corresponds to XGA
-				// register BEE8h.
-
-				// control1 bit 7 is SRC_NE. If clear, don't
-				// update if source value == COLOR_CMP. If set,
-				// don't update if source value != COLOR_CMP
-				doit = !!(((srcval == colorcmpdata) ? 0 : 1) ^
-				          ((xga.control1 >> 7u) & 1u));
-			}
+			auto should_mix_and_draw = [&]() {
+				using namespace bit::literals;
+				// Always update if we're not comparing (COLOR_CMP is bit 8)
+				if (bit::cleared(xga.control1, b8)) {
+					return true;
+				}
+				// If the SRC_NE bit (bit 7) is set ...
+				const auto src_ne_enabled = bit::is(xga.control1, b7);
+				const auto colours_match = (srcval == colorcmpdata);
+				// Then update on a match otherwise update if they differ.
+				return src_ne_enabled == colours_match;
```

Comment:
Nice, the original was madness 😅

Reply:
It took me a while to puzzle my way out of that one! Jenga for the mind :-)